### PR TITLE
Supported a list of files in multipart form data

### LINKED ISSUE #631: Limitation for multipart/form-data form
## Feature Request
The current httpClient of bdk doesn't support multiple values for a given form key. 
With the current httpClient of bdk, this works: 
```java
httpClient.formParam("file", new File("path/to/myFile/");
```

But when we keep adding other files like the snippet bellow doesn't work because we keep overriding the map containing the parameters:
```java
httpClient.formParam("file", new File("path/to/myFirstFile/");
httpClient.formParam("file", new File("path/to/mySecondFile/");
```

### Description of Problem:
 I checked the code and it seems like the form params are stored in a map, and the value of "file" will be overridden each time which will prevent us from having a field with many files. The implementation ideally would support adding of multiple files under the same key and use boundary to separate the files under the field **_file_**.

Ideally we want to to be able to add several files in one field of form data in an iterative way, like this:
```java
httpClient.formParam("file", new File("path/to/myFirstFile/");
httpClient.formParam("file", new File("path/to/mySecondFile/");
httpClient.formParam("file", new File("path/to/myThirdFile/");
```

### Potential Solutions:
N/A

### PR DESCRIPTION
**Pull request by symphony-elias**
### Description
Closes #631

### Checklist
- [x] Referenced an issue in the PR title or description
- [x] Filled properly the description and dependencies, if any
- [x] Unit/Integration tests updated or added
- [N/A] Javadoc added or updated
- [N/A] Updated the documentation in [docs folder](../docs)

### CODE REVIEW

File: symphony-bdk-http/symphony-bdk-http-webclient/src/main/java/com/symphony/bdk/http/webclient/ApiClientWebClient.java [hunk 1]
```diff
@@ -246,6 +249,11 @@ private void serializeMultiPartDataEntry(String paramKey, Object paramValue,
     }
   }
 
+  private void addFileToFormParam(String paramKey, File paramValue, MultiValueMap<String, Object> formValueMap) {
+    File file = paramValue;
+    formValueMap.add(paramKey, new FileSystemResource(file));
```
Comment (imadamahmid1) [inline symphony-bdk-http/symphony-bdk-http-webclient/src/main/java/com/symphony/bdk/http/webclient/ApiClientWebClient.java:254]: does this file get stored in the machine?

Comment (symphony-elias) [reply symphony-bdk-http/symphony-bdk-http-webclient/src/main/java/com/symphony/bdk/http/webclient/ApiClientWebClient.java:254]: The file has to be stored on the machine and accessible by the Java process running this. 
I guess your question is: does the file get closed after HTTP request is made. I have to check that.

File: symphony-bdk-http/symphony-bdk-http-api/src/main/java/com/symphony/bdk/http/api/HttpClient.java [hunk 2]
```diff
@@ -413,13 +417,16 @@ public HttpClient build() {
       return this.queryParams;
     }
 
-    public Map<String, Object> appendFormParam(String key, Object value) {
+    public Map<String, List<Object>> appendFormParam(String key, Object value) {
```
Comment (symphony-youri) [inline symphony-bdk-http/symphony-bdk-http-api/src/main/java/com/symphony/bdk/http/api/HttpClient.java:420]: aren't we breaking the API if we change that?
but then it looks like this part is internal only

**Review by symphony-youri — COMMENTED** [review]
🙏 put a description in the commit message to give a bit of context

File: symphony-bdk-http/symphony-bdk-http-api/src/main/java/com/symphony/bdk/http/api/util/ApiUtils.java [hunk 3]
```diff
@@ -52,6 +54,15 @@ public static void addDefaultRootCaCertificates(KeyStore trustStore) throws Gene
     }
   }
 
+  public static boolean isCollectionOfFiles(Object paramValue) {
+    if (!(paramValue instanceof Collection<?>)) {
+      return false;
+    }
+
+    final Collection<?> collection = (Collection<?>) paramValue;
+    return !collection.isEmpty() && collection.stream().findAny().get() instanceof File;
```
Comment (symphony-youri) [inline symphony-bdk-http/symphony-bdk-http-api/src/main/java/com/symphony/bdk/http/api/util/ApiUtils.java:63]: should we check that all elements are files?

File: symphony-bdk-http/symphony-bdk-http-api/src/main/java/com/symphony/bdk/http/api/HttpClient.java (hunk 2 above)
Comment (symphony-elias) [reply symphony-bdk-http/symphony-bdk-http-api/src/main/java/com/symphony/bdk/http/api/HttpClient.java:420]: Yes this is internal only

File: symphony-bdk-http/symphony-bdk-http-jersey2/src/main/java/com/symphony/bdk/http/jersey2/ApiClientJersey2.java [hunk 4]
```diff
@@ -430,18 +432,12 @@ protected boolean isJsonMime(String mime) {
...
-            .size(file.length())
-            .build();
-        final FormDataBodyPart streamPart = new FormDataBodyPart(
-            contentDisposition,
-            file,
-            MediaType.APPLICATION_OCTET_STREAM_TYPE
-        );
-        multiPart = (FormDataMultiPart) multiPart.bodyPart(streamPart);
+        multiPart = buildFileBodyPart(multiPart, param.getKey(), (File) param.getValue());
+      } else if (isCollectionOfFiles(param.getValue())) {
```
[the remark this answers is not in the document]
Comment (symphony-elias) [reply symphony-bdk-http/symphony-bdk-http-jersey2/src/main/java/com/symphony/bdk/http/jersey2/ApiClientJersey2.java:436]: If it is a collection of something else, we will go to the else statement:
`multiPart = multiPart.field(param.getKey(), this.parameterToString(param.getValue()));`

**Review by symphony-youri — APPROVED** [review]

**Review by symphony-mariacristina — APPROVED** [review]

**Pull request merged**